### DISCUSSION TITLE
How to fix code in NAPI?

### DISCUSSION THREAD

**Discussion by Lodin** [ast-grep/ast-grep#478] [Q&A] [answered]
Trying to use NAPI for replacing some text but didn't find any function for it. How can I replace nodes using NAPI?

**Comment by HerringtonDarkholme** [comment, marked as answer]
Hi, NAPI fix is not shipped now because most JS code needs sourcemap support, which is not supported at the moment.

To replace code, using `magic-string` is recommended for JS source code manipulation. You may need to write replacement string manually, though.

See https://github.com/vue-vine/vue-vine/blob/0d302140f2a0a606cca62a130082952ec183628a/packages/compiler/src/transform.ts#L32-L37 as an example.

**Reply by Lodin** [reply]
Ah, I see, thanks!

**Comment by juanpprieto** [comment]
I think it would be incredibly useful to be able to apply a `fix` programmatically with napi like with are able to do in rules. rules can be limiting if the `fix` is a dynamic value coming from an external source/process.

I don't think source maps are always a requirement. In my use case, I am attempting to pre-process/transform code before the compilation step so, source maps generation is left to the compiler.

Thanks for this great tool. Love it.
